File: core/hooks/src/main/scala/me/seroperson/reload/live/hook/RuntimeShutdownHook.scala
package me.seroperson.reload.live.hook

import me.seroperson.reload.live.build.BuildLogger
import me.seroperson.reload.live.settings.DevServerSettings

/** Generic runtime shutdown hook that runs application shutdown hooks.
  *
  * This hook is designed to gracefully shut down applications by running all
  * registered shutdown hooks. It's a fallback option that works with any Java
  * application regardless of the specific framework used.
  */
class RuntimeShutdownHook extends Hook {

  override def description: String = "Shutdown a generic application"

  override def isAvailable: Boolean =
    true

  override def hook(settings: DevServerSettings, logger: BuildLogger): Unit = {
    ReflectionUtils.runApplicationShutdownHooks()
  }

}

File: core/build-link/src/main/java/me/seroperson/reload/live/hook/RuntimeShutdownHook.java
package me.seroperson.reload.live.hook;

import me.seroperson.reload.live.build.BuildLogger;
import me.seroperson.reload.live.settings.DevServerSettings;

/**
 * Generic runtime shutdown hook that runs application shutdown hooks.
 *
 * <p>This hook is designed to gracefully shut down applications by running all registered shutdown
 * hooks. It's a fallback option that works with any Java application regardless of the specific
 * framework used.
 */
public class RuntimeShutdownHook implements Hook {

  @Override
  public String description() {
    return "Shutdown a generic application";
  }

  @Override
  public boolean isAvailable() {
    return true;
  }

  @Override
  public void hook(Thread th, ClassLoader cl, DevServerSettings settings, BuildLogger logger) {
    ReflectionUtils.runApplicationShutdownHooks(logger);
  }
}
